credentials/xds/xds.go

```diff
@@ -0,0 +1,339 @@
+	}
+	return val
+}
+
+// ClientHandshake performs the TLS handshake on the client-side.
+//
+// It looks for the presence of a HandshakeInfo value in the passed in context
+// (added using a call to NewContextWithHandshakeInfo()), and retrieves identity
+// and root certificates from there. It also retrieves a list of acceptable SANs
+// and uses a custom verification function to validate the certificate presented
+// by the peer. It uses fallback credentials if no HandshakeInfo is present in
+// the passed in context.
+func (c *credsImpl) ClientHandshake(ctx context.Context, authority string, rawConn net.Conn) (net.Conn, credentials.AuthInfo, error) {
+	if !c.isClient {
+		return nil, nil, errors.New("ClientHandshake() is not supported for server credentials")
+	}
+
+	chi := handshakeInfoFromCtx(ctx)
+	if chi == nil {
+		// A missing handshake info in the provided context could mean either
+		// the user did not specify an `xds` scheme in their dial target or that
+		// the xDS server did not provide any security configuration. In both of
+		// these cases, we use the fallback credentials specified by the user.
+		return c.fallback.ClientHandshake(ctx, authority, rawConn)
+	}
+	if err := chi.validate(c.isClient); err != nil {
+		return nil, nil, err
+	}
+
+	// We build the tls.Config with the following values
+	// 1. Root certificate as returned by the root provider.
+	// 2. Identity certificate as returned by the identity provider. This may be
+	//    empty on the client side, if the client is not doing mTLS.
+	// 3. InsecureSkipVerify to true. Certificates used in Mesh environments
+	//    usually contains the identity of the workload presenting the
+	//    certificate as a SAN (instead of a hostname in the CommonName field).
+	//    This means that normal certificate verification as done by the
+	//    standard library will fail.
+	// 4. Key usage to match whether client/server usage.
+	// 5. A `VerifyPeerCertificate` function which performs normal peer
+	// 	  cert verification using configured roots, and the custom SAN checks.
+	var certs []tls.Certificate
+	var roots *x509.CertPool
+	err := func() error {
+		// We use this anonymous function trick to be able to defer the unlock.
+		chi.mu.Lock()
+		defer chi.mu.Unlock()
+
+		if chi.rootProvider != nil {
+			km, err := chi.rootProvider.KeyMaterial(ctx)
+			if err != nil {
+				return fmt.Errorf("fetching root certificates failed: %v", err)
+			}
+			roots = km.Roots
+		}
+		if chi.identityProvider != nil {
+			km, err := chi.identityProvider.KeyMaterial(ctx)
+			if err != nil {
+				return fmt.Errorf("fetching identity certificates failed: %v", err)
+			}
+			certs = km.Certs
+		}
+		return nil
+	}()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	cfg := &tls.Config{
+		Certificates:       certs,
+		InsecureSkipVerify: true,
+	}
+	var keyUsages []x509.ExtKeyUsage
+	if c.isClient {
+		cfg.RootCAs = roots
+		keyUsages = []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth}
+	} else {
```

Comment:
Maybe we can safely remove this condition, since we are sure this is on the client side?

Reply:
Done. Thanks.